File: 0649-dota2-senate/0649-dota2-senate.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        queue<int> q1, q2;
        int n = senate.size();
        for (int i = 0; i < n; i++) {
            (senate[i] == 'R')? q1.push(i): q2.push(i);
        }
        while (!q1.empty() && !q2.empty()) {
            int rIndex = q1.front(), dIndex = q2.front();
            q1.pop();
            q2.pop();
            (rIndex < dIndex)? q1.push(rIndex + n): q2.push(dIndex + n);
        }
        return (q1.size() > q2.size())? "Radiant": "Dire";
    }
};

File: 0649-dota2-senate/0649-dota2-senate.java
class Solution {
    public String predictPartyVictory(String senate) {
        int len = senate.length();
        Queue<Integer> queue1 = new LinkedList<>();
        Queue<Integer> queue2 = new LinkedList<>();
        for (int i = 0; i < len; i++) {
            if (senate.charAt(i) == 'R') {
                queue1.offer(i);
            } else {
                queue2.offer(i);
            }
        }
        while (!queue1.isEmpty() && !queue2.isEmpty()) {
            int radiantIndex = queue1.poll();
            int direIndex = queue2.poll();
            if (radiantIndex < direIndex) {
                queue1.offer(radiantIndex + len);
            } else {
                queue2.offer(direIndex + len);
            }
        }
        return (queue2.isEmpty()) ? "Radiant" : "Dire";
    }
}